### PR TITLE
scripts: Avoid recomputing the validator scripts

### PR DESCRIPTION
**Pull request by j-mueller**
* Change the `ScriptInstance` type to include the validator script and the script address
* Slightly rearrange things in the multi sig state machine contract to increase sharing

Before:
grafik

After: 
grafik

### CODE REVIEW

**Comment by j-mueller** [comment]
The futures contract is still relatively slow. It needs to be changed to use the new state machine lib, but I'll do that in another PR

File: plutus-wallet-api/ledger/Ledger/Typed/Scripts.hs [hunk 1]
```diff
@@ -33,20 +45,49 @@ type ValidatorType (a :: Type) = DataType a -> RedeemerType a -> Validation.Pend
...
 
 -- | The type of a connection.
-data ScriptInstance (a :: Type) where
+data ScriptConnection (a :: Type) where
     Validator
         :: ScriptType a
         => CompiledCode (ValidatorType a)
         -> CompiledCode (ValidatorType a -> WrappedValidatorType)
-        -> ScriptInstance a
+        -> ScriptConnection a
```
Comment (michaelpj) [inline plutus-wallet-api/ledger/Ledger/Typed/Scripts.hs]: I was thinking to just add more fields to this, and then have a smart constructor. This works too, I'm just not wild about `ScriptConnection` as a name...

**Comment by kwxm** [comment]
That's more like it!  Well done.

File: plutus-wallet-api/ledger/Ledger/Typed/Scripts.hs (hunk 1 above)
Comment (j-mueller) [reply plutus-wallet-api/ledger/Ledger/Typed/Scripts.hs]: I deleted the `ScriptConnection` type. We don't actually use the `CompiledCode` values for anything other than getting the `ValidatorScript` and the `Address`.

Comment (michaelpj) [reply plutus-wallet-api/ledger/Ledger/Typed/Scripts.hs]: Yeah, it's really only an intermediate stage.

**Pull request merged**